splash/qtrender_lua.py

```diff
@@ -487,6 +491,75 @@ def error(error_info):
         ))
         return cmd
 
+    @command(async=True, sets_callback=True, decode_arguments=False)
+    def with_timeout(self, timeout, callback):
+        if timeout is None:
+            ScriptError({
+                "argument": "timeout",
+                "message": "splash:with_timeout timeout must be specified",
+                "splash_method": "with_timeout",
+            })
+        if not isinstance(timeout, (float, int)):
+            raise ScriptError({
+                "argument": "timeout",
+                "message": "splash:with_timeout timeout must be a number",
+                "splash_method": "with_timeout",
+            })
+
+        timeout = int(float(timeout) * 1000)
+
+        if timeout < 0:
+            raise ScriptError({
+                "argument": "timeout",
+                "message": "splash:with_timeout timeout must be >= 0",
+                "splash_method": "with_timeout",
+            })
+
+        if lupa.lua_type(callback) != 'function':
+            raise ScriptError({
+                "argument": "callback",
+                "message": "splash:with_timeout callback is not a function",
+                "splash_method": "with_timeout",
+            })
+
+        qtimer = QTimer(self.tab)
+        qtimer.setSingleShot(True)
+
+        def timer_callback():
+            run_coro.runner.stop()
+            cmd.return_result(False, "timeout_over")
+
+        qtimer.timeout.connect(timer_callback)
+
+        def coro_success(result):
+            if not qtimer.isActive():
+                return
+
+            qtimer.stop()
+            cmd.return_result(True, result)
+
+        def coro_error(ex):
+            if not qtimer.isActive():
+                return
+
+            qtimer.stop()
+
+            # TODO: make error more verbose (e.g. add line number)
+            cmd.return_result(False, str(ex.args[0]))
+
+        run_coro = self.get_coroutine_run_func(
+            "splash:with_timeout", callback, coro_success, coro_error)
+
+        def start():
+            qtimer.start(timeout)
+            run_coro.runner = run_coro()
```

Comment:
Passing a variable as a new attribute looks a bit odd; there can be other ways to write this code (e.g. define a callback inside this function and pass variable in a closure), but I'm not sure they are better though. But if you have ideas on how to clean it up then please do it :)